components/api/src/error.rs

```diff
@@ -14,49 +14,116 @@
-    #[fail(display = "Not found: {}", _0)]
-    NotFound(String),
-
-    /// Internal server error. This type can return any internal server error to the user.
-    #[fail(display = "Internal server error: {}", _0)]
-    InternalError(failure::Error),
-
-    /// Unauthorized error. This error occurs when the request lacks valid
-    /// authentication credentials.
-    #[fail(display = "Unauthorized")]
-    Unauthorized,
+use serde::{Deserialize, Serialize};
+
+use std::fmt;
+
+/// API HTTP error struct.
+#[derive(Fail, Debug, PartialEq)]
+pub struct Error {
+    /// HTTP error code.
+    pub http_code: HttpStatusCode,
+    /// API error body.
+    pub body: ErrorBody,
+    /// Additional HTTP headers.
+    pub headers: HeaderMap,
+}
+
+#[derive(Debug, Serialize, Deserialize, Default, PartialEq)]
+pub struct ErrorBody {
+    /// A URI reference to the documentation or possible solutions for the problem.
+    #[serde(rename = "type", default, skip_serializing_if = "String::is_empty")]
+    pub docs_uri: String,
+    /// Short description of the error.
+    #[serde(default, skip_serializing_if = "String::is_empty")]
+    pub title: String,
+    /// Detailed description of the error.
+    #[serde(default, skip_serializing_if = "String::is_empty")]
+    pub detail: String,
+    /// Source of the error.
+    #[serde(default, skip_serializing_if = "String::is_empty")]
+    pub source: String,
+    /// Internal error code.
+    #[serde(default, skip_serializing_if = "Option::is_none")]
+    pub error_code: Option<u8>,
+}
+
+impl fmt::Display for Error {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "{}: {}", self.body.title, self.body.detail)
+    }
+}
+
+impl Error {
+    /// Builds a ApiError with the given `http_code`.
+    pub fn new(http_code: HttpStatusCode) -> Self {
+        Self {
+            http_code,
+            body: ErrorBody::default(),
+            headers: HeaderMap::new(),
+        }
+    }
+
+    /// Sets `docs_uri` of an error.
+    pub fn docs_uri(mut self, docs_uri: impl Into<String>) -> Self {
+        self.body.docs_uri = docs_uri.into();
+        self
+    }
+
+    /// Sets `title` of an error.
+    pub fn title(mut self, title: impl Into<String>) -> Self {
+        self.body.title = title.into();
+        self
+    }
+
+    /// Sets `detail` of an error.
+    pub fn detail(mut self, detail: impl Into<String>) -> Self {
+        self.body.detail = detail.into();
+        self
+    }
+
+    /// Sets `source` of an error.
+    #[doc(hidden)]
+    pub fn source(mut self, source: impl Into<String>) -> Self {
+        self.body.source = source.into();
+        self
+    }
+
+    /// Sets `error_code` of an error.
+    pub fn error_code(mut self, error_code: u8) -> Self {
+        self.body.error_code = Some(error_code);
+        self
+    }
+
+    /// Adds HTTP header, which will be added in `HttpResponse`
+    #[doc(hidden)]
+    pub fn header(mut self, key: HeaderName, value: &str) -> Self {
```

Comment:
Does this method need to be public? As far as I see, it's only used in this file. I'd prefer to include more methods later as they become necessary.